src/eager.rs

```diff
@@ -12,3 +12,181 @@ pub use tensor_handle::*;
 mod op;
 
 pub use op::raw_ops;
+
+use crate::{Result, Tensor, TensorType};
+
+#[cfg(feature = "ndarray")]
+use ndarray::{ArrayBase, Data, Dimension};
+
+/// A helper trait to convert a Tensor or some other types into a TensorHandle
+/// for use in eager execution.
+pub trait ToTensorHandle<'a> {
+    /// Convert a Tensor or values into a new TensorHandle.
+    ///
+    /// _Warning_ : This function may create multiple handles to the same
+    /// underlying tensor. Users should be careful not to modify the tensor
+    /// after converting it to a TensorHandle. Also, users should be careful
+    /// not to modify the Tensor generated from the TensorHandle's `resolve`
+    /// method.
```

Comment:
I should have caught this sooner, but this doesn't just have the potential to cause user confusion and bugs; it could actually invoke undefined behavior: https://doc.rust-lang.org/reference/behavior-considered-undefined.html.  (It falls under "Breaking the pointer aliasing rules".)  At least, if the data is modified while it's held behind a reference in Rust elsewhere, it's definitely undefined behavior.  With the FFI involved, it's less clear.  I have no idea if e.g. modifying data through a `&mut` and reading it in C++ while the reference is still live causes undefined behavior or not, although it probably is.

The eager mode support needs to guarantee through some mechanism that no safe code can break the pointer aliasing rules.  Unfortunately, the C API doesn't make that easy, since it only says `TFE_TensorHandleResolve` *might* alias internal memory, and there's no clear way to find out what it is or isn't aliasing.  We could do a defensive copy, although that would be less efficient, of course.  Another idea would be to create something like a `ReadonlyTensor` type which looks like `Tensor` but only supports readonly operations.

We also need to take care of `TensorHandle::new`.  It probably has the same fixes (either copy, i.e. take a `Tensor` instead of a `&Tensor`, or pass a `&ReadonlyTensor`).

It also occurs to me that `TensorHandle::new` has lifetime issues:

```rust
let ctx: Context = ...;
let handle = {
  let tensor: Tensor<f32> = ...;
  TensorHandle::new(&ctx, &tensor)?
};
// At this point, the tensor has deallocated its memory, but we still have a handle.
```

That could be fixed by adding a constraint so that the handle outlives both the context and the tensor, e.g.

```rust
impl<'a> TensorHandle<'a> {
    pub fn new<T: TensorType, 'b: 'a, 'c: 'a>(_ctx: &'b Context, t: &'c Tensor<T>) -> Result<TensorHandle<'a>> {
```

An alternative to all of the above is to mark all of these methods as unsafe and require the client to manage this somehow.  That's not a very user-friendly interface, though.

Reply:
> I should have caught this sooner, but this doesn't just have the potential to cause user confusion and bugs; it could actually invoke undefined behavior: https://doc.rust-lang.org/reference/behavior-considered-undefined.html. (It falls under "Breaking the pointer aliasing rules".) At least, if the data is modified while it's held behind a reference in Rust elsewhere, it's definitely undefined behavior. With the FFI involved, it's less clear. I have no idea if e.g. modifying data through a `&mut` and reading it in C++ while the reference is still live causes undefined behavior or not, although it probably is.

I have a similar concern, for example, the following code seems to cause non-mutable memory to be modified outside of that variable. I added a unittest to ensure the eager mode API does not break the Rust's memory design. I think this is partially related to what you pointed. Currently, this unittest will fail:

https://github.com/tensorflow/rust/blob/05c49af1b4f3548b765c026d0ddca11662db9e4f/src/eager.rs#L101-L122

> The eager mode support needs to guarantee through some mechanism that no safe code can break the pointer aliasing rules. Unfortunately, the C API doesn't make that easy, since it only says `TFE_TensorHandleResolve` _might_ alias internal memory, and there's no clear way to find out what it is or isn't aliasing. We could do a defensive copy, although that would be less efficient, of course. Another idea would be to create something like a `ReadonlyTensor` type which looks like `Tensor` but only supports readonly operations.

It seems a good idea to create a ReadonlyTensor, but it was not immediately clear to what extent it should be the ReadonlyTensor for the eager execution. I will think about it some more in conjunction with the above problem.

> We also need to take care of `TensorHandle::new`. It probably has the same fixes (either copy, i.e. take a `Tensor` instead of a `&Tensor`, or pass a `&ReadonlyTensor`).
> 
> It also occurs to me that `TensorHandle::new` has lifetime issues:
> 
> ```rust
> let ctx: Context = ...;
> let handle = {
>   let tensor: Tensor<f32> = ...;
>   TensorHandle::new(&ctx, &tensor)?
> };
> // At this point, the tensor has deallocated its memory, but we still have a handle.
> ```

If we can rely on the implementation detail of the TensorFlow's Tensor class, this might be fine because each Tensor is reference counted, and creating new TensorHandle increments the count. So destructing Tensor while leaving only its handles may be fine, although we cannot read out this point from the `eager/c_api.h`. I will consider these points a little more as well, including the following points.

> 
> That could be fixed by adding a constraint so that the handle outlives both the context and the tensor, e.g.
> 
> ```rust
> impl<'a> TensorHandle<'a> {
>     pub fn new<T: TensorType, 'b: 'a, 'c: 'a>(_ctx: &'b Context, t: &'c Tensor<T>) -> Result<TensorHandle<'a>> {
> ```
> 
> An alternative to all of the above is to mark all of these methods as unsafe and require the client to manage this somehow. That's not a very user-friendly interface, though.

Reply:
I introduced ReadonlyTensor at a4111df. The ReadonlyTensor is a Tensor that does not have mutation operations of the underlying memory. The conversion between Tensor and ReadonlyTensor is performed by moving the ownership of the underlying buffers.

__Conversion steps__

* `Tensor` -(`freeze`)-> `ReadonlyTensor` -(`to_handle`)-> `TensorHandle`
* `TensorHandle` -(`resolve`)-> `ReadonlyTensor` -(`into_tensor`)-> `Tensor`

The unsafe point is a conversion from ReadonlyTensor to Tensor, which will result in "Breaking the pointer aliasing rules" when modified. So it is marked as unsafe (the `into_tensor` method).

Ideally, it would be more useful if the ReadonlyTensor could be fed into the `SessionRunArgs` directly, but it is not done since it needs more work (maybe separating Tensor's write/read interface).

Reply:
> > We also need to take care of `TensorHandle::new`. It probably has the same fixes (either copy, i.e. take a `Tensor` instead of a `&Tensor`, or pass a `&ReadonlyTensor`).
> > It also occurs to me that `TensorHandle::new` has lifetime issues:
> > ```rust
> > let ctx: Context = ...;
> > let handle = {
> >   let tensor: Tensor<f32> = ...;
> >   TensorHandle::new(&ctx, &tensor)?
> > };
> > // At this point, the tensor has deallocated its memory, but we still have a handle.
> > ```
> 
> If we can rely on the implementation detail of the TensorFlow's Tensor class, this might be fine because each Tensor is reference counted, and creating new TensorHandle increments the count. So destructing Tensor while leaving only its handles may be fine, although we cannot read out this point from the `eager/c_api.h`. I will consider these points a little more as well, including the following points.
> 

I checked this point in [the unit test code](https://github.com/tensorflow/tensorflow/blob/c256c071bb26e1e13b4666d1b3e229e110bc914a/tensorflow/c/eager/c_api_test_util.cc#L83-L95). It seems that deleting TF_Tensor while leaving the corresponding TFE_TensorHandle is the valid use case.

```c++
TFE_TensorHandle* DoubleTestMatrixTensorHandle(TFE_Context* ctx) {
  int64_t dims[] = {2, 2};
  double data[] = {1.0, 2.0, 3.0, 4.0};
  TF_Status* status = TF_NewStatus();
  TF_Tensor* t = TFE_AllocateHostTensor(ctx, TF_DOUBLE, &dims[0],
                                        sizeof(dims) / sizeof(int64_t), status);
  memcpy(TF_TensorData(t), &data[0], TF_TensorByteSize(t));
  TFE_TensorHandle* th = TFE_NewTensorHandleFromTensor(ctx, t, status);
  CHECK_EQ(TF_OK, TF_GetCode(status)) << TF_Message(status);
  TF_DeleteTensor(t);
  TF_DeleteStatus(status);
  return th;
}
```
I added this information in the TensorHandle's docs.

Reply:
Interesting.  I didn't realize they were using reference counting.  We may also need a safe (i.e. not `unsafe`) function to copy a `ReadonlyTensor` to a new `Tensor` and a trait to abstract the common behavior for `Tensor` and `ReadonlyTensor`, but we can always add those later as needed.  This looks good for now.